bundle/config/mutator/python/apply_python_mutator.go

```diff
@@ -0,0 +1,190 @@
+	}
+
+	logWriter := newLogWriter(ctx, "stderr: ")
+
+	stdout, err := process.Background(
+		ctx,
+		args,
+		process.WithDir(rootPath),
+		process.WithStderrWriter(logWriter),
+		process.WithStdinReader(bytes.NewBuffer(rootConfigJson)),
+	)
+
+	if err != nil {
+		return dyn.InvalidValue, fmt.Errorf("python mutator process failed: %w", err)
+	}
+
+	// we need absolute path, or because later parts of pipeline assume all paths are absolute
+	// and this file will be used as location
+	virtualPath, err := filepath.Abs(filepath.Join(rootPath, "__generated__.yml"))
+
+	if err != nil {
+		return dyn.InvalidValue, fmt.Errorf("failed to get absolute path: %w", err)
+	}
+
+	generated, err := yamlloader.LoadYAML(virtualPath, bytes.NewReader([]byte(stdout)))
+
+	if err != nil {
+		return dyn.InvalidValue, fmt.Errorf("failed to parse Python mutator output: %w", err)
+	}
+
+	normalized, diagnostic := convert.Normalize(config.Root{}, generated)
+
+	if diagnostic.Error() != nil {
+		return dyn.InvalidValue, fmt.Errorf("failed to normalize Python mutator output: %w", diagnostic.Error())
+	}
+
+	return normalized, nil
+}
+func createOverrideVisitor(ctx context.Context, phase phase) merge.OverrideVisitor {
+	jobsPath := dyn.NewPath(dyn.Key("resources"), dyn.Key("jobs"))
+
+	return merge.OverrideVisitor{
+		VisitDelete: func(valuePath dyn.Path, left dyn.Value) error {
+			if !valuePath.HasPrefix(jobsPath) {
+				return fmt.Errorf("unexpected change at '%s' (delete)", valuePath.String())
+			}
+
+			deleteResource := len(valuePath) == len(jobsPath)+1
+
+			if phase == ApplyPythonMutatorPhasePreInit {
+				return fmt.Errorf("unexpected change at '%s' (delete)", valuePath.String())
+			} else if deleteResource && phase == ApplyPythonMutatorPhaseInit {
+				return fmt.Errorf("unexpected change at '%s' (delete)", valuePath.String())
+			}
+
+			log.Debugf(ctx, "Delete value at '%s'", valuePath.String())
+
+			return nil
+		},
+		VisitInsert: func(valuePath dyn.Path, right dyn.Value) (dyn.Value, error) {
+			if !valuePath.HasPrefix(jobsPath) {
+				return dyn.InvalidValue, fmt.Errorf("unexpected change at '%s' (insert)", valuePath.String())
+			}
+
+			insertResource := len(valuePath) == len(jobsPath)+1
+
+			if phase == ApplyPythonMutatorPhasePreInit && !insertResource {
+				return dyn.InvalidValue, fmt.Errorf("unexpected change at '%s' (insert)", valuePath.String())
+			}
+
+			log.Debugf(ctx, "Insert value at '%s'", valuePath.String())
+
+			return right, nil
+		},
+		VisitUpdate: func(valuePath dyn.Path, left dyn.Value, right dyn.Value) (dyn.Value, error) {
+			if !valuePath.HasPrefix(jobsPath) {
+				return dyn.InvalidValue, fmt.Errorf("unexpected change at '%s' (update)", valuePath.String())
+			}
+
+			if phase == ApplyPythonMutatorPhasePreInit {
+				return dyn.InvalidValue, fmt.Errorf("unexpected change at '%s' (update)", valuePath.String())
+			}
+
+			log.Debugf(ctx, "Update value at '%s'", valuePath.String())
+
+			return right, nil
+		},
+	}
```

Comment:
Given we only have 2 phases, I think it's OK to create one visitor for the pre-init phase, and another one for the other phase to keep things more clear. The branching inside these handlers is somewhat difficult to untangle.